### ISSUE TITLE
Generalize named dimensions to handle indirect access

### ISSUE THREAD

**Issue by tardieu** [torch-spyre/torch-spyre#2433]
### What feature would you like to be added?

Named tensor dimension propagation should support indirect access as in:
```python
import torch

from torch_spyre._inductor import spyre_hint
from torch_spyre._inductor.propagate_named_dims import (
    declare_tensor_dim,
    name_tensor_dims,
)

N = 16384  # cache size
S = 1024  # sequence length


# access random fp16 sticks
def make_random_index_tensor():
    t = torch.empty(S, dtype=torch.int32)
    sticks = torch.randint(0, N // 64, (S // 64,))
    for i in range(S // 64):
        base = sticks[i].item()
        t[i * 64 : (i + 1) * 64] = torch.arange(base * 64, base * 64 + 64)
    return t


x = torch.rand(N, dtype=torch.float16).to("spyre")
i = make_random_index_tensor().to("spyre")

declare_tensor_dim("S", S)

name_tensor_dims(i, ["S"])


def kernel(x, i):
    with spyre_hint(num_tiles_per_dim={"S": 8}):
        return x[i].exp()


print(torch.compile(kernel)(x, i))
``` 

### Why is this feature needed?

To support working set reduction for paged attention.

### Anything else we need to know?

The propagation should result in the `exp` computation being split 8 ways by splitting the index tensor 8 ways.

**Comment by marnold-ibm** [comment]
From what I can tell, the existing code is working correctly for this example.

**With `name_tensor_dims(i, ["S"])`**

```
  [INFO] [propagate_named_dims]     input arg1_1: named_dims=['S']  host_size=[1024]  index=d0  ranges={d0: 1024}
  [INFO] [propagate_named_dims]     input arg0_1: named_dims=[]  host_size=[16384]  index=tmp0  ranges={}
  [INFO] [propagate_named_dims]     loop vars:
  [INFO] [propagate_named_dims]       d0: range=1024  named_dim(s)=['S']  declared=['S=1024']
  [INFO] [propagate_named_dims]     output: (buf0) named_dims=['S']
  [INFO] [propagate_named_dims]   ['S']  range=1024  split_count=8  -> 128 per tile  dim_index=0
  [coarse_tile._stamp_group] op=op0  op_hints=[(1, 0, 8)]  hint_id_to_dim_index={1: 0}  ranges_before=[1024]
  [coarse_tile._stamp_group]   level hint_id=1 count=8  spec_dims=[0]  dim_index=0  effective=[0]
  [coarse_tile._stamp_group]   ranges_after=[128]  op_tiled_dims=[[0]]
```
  - arg1_1 is i, accessed at index=d0 — d0 steps linearly through i
  - d0 maps to named dim S=1024
  - S is tiled by 8 → 128 per tile, applied to dim_index=0
  - ranges_before=[1024] → ranges_after=[128] confirms the division
  - Each coarse tile processes 128 consecutive elements of i: tile 0 = i[0..127], tile 1 = i[128..255], etc.

-------

**With `name_tensor_dims(x, ["S"])`**

```
  [INFO] [propagate_named_dims]     input arg1_1: named_dims=[]  host_size=[1024]  index=d0  ranges={d0: 1024}
  [INFO] [propagate_named_dims]     input arg0_1: named_dims=['S']  host_size=[16384]  index=tmp0  ranges={}
  [INFO] [propagate_named_dims]       d0: range=1024  named_dim(s)=['_untracked_1024']  declared=['_untracked_1024=1024']
  [INFO] [propagate_named_dims]     output: (buf0) named_dims=['_untracked_1024']
  [INFO] [propagate_named_dims]   ['S']  range=0  split_count=8  -> ? per tile  dim_index=None
  [INFO] [temp_passes] coarse_tile_groups: 0 group(s) formed
```
  - arg0_1 is x, accessed at index=tmp0 with ranges={} — no loop variable, it's indirect
  - S on x cannot match any loop var, so d0 falls back to _untracked_1024
  - dim_index=None and range=0 — system correctly found no loop var for S
  - coarse_tile_groups: 0 — no tiling attempted, correct
  - ranges=[1024] unchanged — op runs flat, no coarse loop formed

**Comment by tardieu** [comment]
I see a logging issue for the same example:
```
[INFO] [propagate_named_dims]     input arg1_1: named_dims=?  host_size=[1024]  index=d0  ranges={d0: 1024}
[INFO] [propagate_named_dims]     input arg0_1: named_dims=?  host_size=[16384]  index=tmp0  ranges={}
```
This will be fixed in #2457 (which is assumed in the previous comment).

I suggest we close this issue after merging #2457.

**Comment by marnold-ibm** [comment]
#2457 is merged, so closing.
